readme.md
# 5b

5b (which stands for base base base base BASE) is a command line tool aimed at developers to make it easier to convert between bases and other types of encodings.

No more learning or having to look up conversion codes, just enter the desired string, select how you want 5b to interpret the string, then 5b will convert that value to as many types as it can.

Supported types include: 

- Bases 2, 8, 10 and 16
- 16, 32 and 64-bit float
- 8, 16, 32, 64 and 128-bit signed and unsigned integers
- 32 and 64-bit unix time
- Unicode 8
- Ipv4 and Ipv6 addresses
- Base 64, 85 and 91
- Byte list
- UUID
- Escaped string
- URL percent encode/decode


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]
### To Do
- ToIR implementations:
  - File path (load bytes from file, implement as a command line option)  
- FromIR implementations:
  - Disassembly (16, 32 and 64-bit)
  - File path (save bytes from file, implement as a command line option, with optional prompt to confirm the save)
- Add command line options for
  - Hex case formatting
  - ascii table
  - Forcing little/big endianness (instead of prompting user) when converting to bytes
- Ignore whitespace with certain types (numbers)
- Include some example usage in readme (escaping backslashes for regexes, identifying invalid unicode characters, basic conversions and some esoteric conversions)
- Pipe input to and from 5b
- Get away from `colour` and use the `ansi_term` crate instead
- Add precompiled binary to sourceforge and add link to readme
- Upgrade readme (build instructions, links to precompiled binaries, example usage, explination of  benefits of 5b)

- Big three to focus on:
  - Change colouring to `ansi_term`
  - Sort out colouring, formatting and presentation
  - Include command options
  - Improve readme

### Unfinished Ideas
- Not sure about disassembly. But if we want to use it, [this](https://crates.io/crates/iced-x86) seems to be the solution 

## [0.1.5] - 2021-07-03
### Added
- When the selected type is `Dual` endianness, we ask the user which (little or big) endianness they would like to interpret the data as
- `FixedInt` now converts via `Base2_16` so we accept base 2, 8, 10 and 16 for the primitive integers 
  - This only works when numbers are prefixed (0x, 0o and 0b) otherwise it is treated as decimal.
  - This means that non-prefixed binary numbers will assumed to be decimal by `FixedInt` and non-prefixed octal or hex numbers will be ignored by `FixedInt`
  - They will still be picked up by `Base2_16` though
- `FixedFloat` now displays floats using scientific notation and SI unit notation
- `Endian` trait representing `Endianness` for types 

### Fixed
- We now check for invalid dates (out of range usually) and returns an error string instead of the formatted date
- No longer panics when an empty string is read via `read_without_newline`

### Removed
- `Hash` object removed as this is a bytes to bytes object (at the moment we only deal with string to bytes or bytes to string objects)

## [0.1.4] - 2021-06-26
### Added
- Unicode names which allows the user to 
  - Convert a unicode name into a character
  - Convert a string into a list of unicode names
- Added `common::Colour` to represent colours of various bit depths. Supplies some functions to convert to 24-bit colour  
- `FromIR` implementation for colours supporting
  - 8, 16, 24 and 32-bit colours
  - 8-bit colour, greyscale and terminal formats  
  - RGB and HSL values for 24-bit
- `ToIR` implementation of colours supporting html colour format (#RRGGBB)

### Changed
- `FromIR::encode` now returns `ANSIGenericString<str>` which allows styling of terminal strings
- Showing dual endianness now only colours the brackets magenta to allow styled text

## [0.1.3] - 2021-06-12
### Added
- Mantissa/Exponent form for `FixedFloat` added
- Endianness
  - `ToIr`: Always store data as little endian
  - `FromIr`: Two options, display the default endianness (chosen by `FromIr::encode`) or both little and big if no default can be chosen (primitive integers and float for example)
  - Endianness is implemented by
    - Adding a function to `FromIR` that identifies if a type has default or dual endianness
    - The main control flow by displaying a variants on their own (default) or two values for each variant, one little and another big endian (dual)

### Fixed
- Escape sequence code now accepts \fff as a valid python sequence as well as a C sequence

## [0.1.2] - 2021-06-05
### Added
- Byte list now shows the number of bytes 
- Variants for C and Python escape sequences
- Python style Named unicode characters added to the escape sequence type (\N)
- Unicode escape characters (\U and \u)

### Changes
- FromIT typo corrected to FromIR
- EscapedString renamed to EscapeSequence

### Fixed
- When checking for prefixes ('0x', '0o', etc.) we now test for the size before we take a slice

## [0.1.1] - 2021-06-02
### Added
- Base conversion has been shortened to base 2, 8, 10 and 16
- Endianness of Ipv4 and Ipv6 is fixed
- Colour added to terminal output
- Prefixes 0x, 0b and 0o are now supported
- URL decoding and encoding added
- `FromIT::variants` and `ToIR::identify` return an Option<Vec<_>> and return None if no variants match to avoid unnecessary allocations 
- Warnings cleared

## [0.1.0] - 2021-06-01
### Added
- Readme
- This changelog
- License
- Basic command line interface
- Escaped string
- UUID FromIR
- Removed Hash object

src/main.rs
extern crate lazy_static;

mod common;
mod fromir;
mod toir;
mod escape;
mod endian;

use fromir::FromIR;
use toir::ToIR;
use crate::common::{Variant, FixedInt, FixedFloat, Base2_16, DateTime, Unicode8, IpV4, IpV6, Base91, Base64, Base85, ByteList, UUID, EscapedString, UrlEncode, UrlDecode, UnicodeNames};

use colour::{blue, yellow, green, magenta};
use crate::endian::{Endianness, Endian};
use ansi_term::{ANSIGenericString};


fn read_without_newline() -> String {
    let mut string = String::new();

    std::io::stdin().read_line(& mut string).expect("Failed to read from stdin");

    if !string.is_empty() {
        if string.bytes().last().unwrap() == 10 {
            string.remove(string.len() - 1);
        }
    }

    if !string.is_empty() {
        if string.bytes().last().unwrap() == 13 {
            string.remove(string.len() - 1);
        }
    }

    string

}

fn main() {

    let to_ir: Vec<(fn(& str) -> Option<Vec<Variant>>, & str, fn(& str, Variant) -> Vec<u8>, fn() -> Endianness)> = vec![
        (IpV4::identify, "Ipv4 address", IpV4::decode, IpV4::endianness),
        (IpV6::identify, "Ipv6 address", IpV6::decode, IpV6::endianness),
        (DateTime::identify, "Unix time", DateTime::decode, DateTime::endianness),
        (FixedFloat::identify, "Floats", FixedFloat::decode, FixedFloat::endianness),
        (UUID::identify, "UUID", UUID::decode, UUID::endianness),
        (FixedInt::identify, "Primitive integers", FixedInt::decode, FixedInt::endianness),
        (Base2_16::identify, "Base 2-16 number", Base2_16::decode, Base2_16::endianness),
        (Base64::identify, "Base64 data", Base64::decode, Base64::endianness),
        (Base85::identify, "Base85 data", Base85::decode, Base85::endianness),
        (Base91::identify, "Base91 data", Base91::decode, Base91::endianness),
        (Unicode8::identify, "Unicode 8 string", Unicode8::decode, Unicode8::endianness),
        (ByteList::identify, "Byte list", ByteList::decode, ByteList::endianness),
        (EscapedString::identify, "Escaped sequence", EscapedString::decode, EscapedString::endianness),
        (UnicodeNames::identify, "Unicode character names", UnicodeNames::decode, UnicodeNames::endianness),
        (common::Colour::identify, "HTML colour", common::Colour::decode, common::Colour::endianness),
    ];

    let from_ir: Vec<(fn(& [u8]) -> Option<Vec<Variant>>, & str, fn(& [u8], Variant) -> ANSIGenericString<str>, fn() -> Endianness)> = vec![
        (IpV4::variants, "Ipv4 address", IpV4::encode, IpV4::endianness),
        (IpV6::variants, "Ipv6 address", IpV6::encode, IpV6::endianness),
        (DateTime::variants, "Unix time", DateTime::encode, DateTime::endianness),
        (FixedFloat::variants, "Floats", FixedFloat::encode, FixedFloat::endianness),
        (UUID::variants, "UUID", UUID::encode, UUID::endianness),
        (FixedInt::variants, "Primitive integers", FixedInt::encode, FixedInt::endianness),
        (Base2_16::variants, "Base 2-16 numbers", Base2_16::encode, Base2_16::endianness),
        (Base64::variants, "Base64 data", Base64::encode, Base64::endianness),
        (Base85::variants, "Base85 data", Base85::encode, Base85::endianness),
        (Base91::variants, "Base91 data", Base91::encode, Base91::endianness),
        (ByteList::variants, "Byte list", ByteList::encode, ByteList::endianness),
        (Unicode8::variants, "Unicode 8 string", Unicode8::encode, Unicode8::endianness),
        (EscapedString::variants, "Escaped sequence", EscapedString::encode, EscapedString::endianness),
        (UrlEncode::variants, "Encoded URL", UrlEncode::encode, UrlEncode::endianness),
        (UrlDecode::variants, "Decoded URL", UrlDecode::encode, UrlDecode::endianness),
        (common::Colour::variants, "Colour", common::Colour::encode, common::Colour::endianness),
    ];



    println!("Please enter the input string:");

    let input = read_without_newline();

    let mut option_map = Vec::new();

    println!("***************");
    println!("Possible types:");
    println!("***************");

    for (identifier, name, decoder, endianness) in to_ir {

        let optional_variants = (identifier)(&input);

        if let Some(variants) = optional_variants {
            println!("{}", name);

            for variant in variants {
                blue!("    {}", option_map.len());
                println!("     {}", variant.0);

                option_map.push((variant, decoder, endianness()));
            }
        }
    }

    println!("Please enter the index of the possible format you would like to use:");

    let option = read_without_newline();

    match option.parse::<usize>() {
        Ok(option) => {

            if option >= option_map.len() {
                panic!("Invalid number (must be between 0 and {}, inclusive)", option_map.len()-1);
            }

            let (variant, decoder, endianness) = option_map[option].clone();

            let (little_endian_ir, big_endian_ir) = {


                let ir = (decoder)(&input, variant);

                let reverse: Vec<_> = ir.iter().rev().map(|x| *x).collect();

                if let Endianness::Dual = endianness {
                    println!("Would you like the result to be interpreted as little (l) or big (b) endian? If you're not sure, choose the default, little");
                    let user_endianness = read_without_newline();

                    match user_endianness.to_lowercase().as_str() {
                        "l" | "little" | "" => {
                            (ir, reverse)
                        }
                        "b" | "big" => {
                            (reverse, ir)
                        }
                        _ => {
                            println!("Invalid endianness ('l' or 'little' for little endianness and 'b' or 'big' for big endianness");
                            return;
                        }
                    }
                } else {
                    (ir, reverse)
                }
            };


            println!("**************");
            println!("Other formats:");
            println!("**************");

            for (variants_function, name, encoder, endianness) in from_ir {

                let endianness = endianness();

                let optional_variants = (variants_function)(&little_endian_ir);

                if let Some(variants) = optional_variants {
                    green!("{}", name);
                    println!();

                    for variant in variants {

                        yellow!("    {}", variant.0);

                        match endianness {
                            Endianness::Default => {
                                println!("    {}", (encoder)(&little_endian_ir, variant.clone()));
                            },
                            Endianness::Dual => {
                                print!("    {} ", (encoder)(&little_endian_ir, variant.clone()));
                                magenta!("(");
                                print!("{}", (encoder)(&big_endian_ir, variant.clone()));
                                magenta!(")");
                                println!();
                            }
                        }
                    }
                }
            }
        }
        Err(_) => {
            println!("Invalid number {:?}", option);
        }
    }


}


src/common.rs
use std::str::FromStr;
use std::convert::TryInto;
use crate::endian::Endianness;
use crate::toir::ToIR;

//Convertable types
pub struct Base2_16; //All numbers between base 2 and 16, each base implemented as a variant
pub struct FixedFloat; //16, 32, 64-bit floats
pub struct DateTime; //32 and 64-bit unix time
pub struct FixedInt; //8-128 bit signed/unsigned integer
pub struct Unicode8;
pub struct IpV4;
pub struct IpV6;
pub struct Base64;
pub struct Base85;
pub struct Base91;
pub struct ByteList;
pub struct UUID;
pub struct EscapedString;
pub struct UrlEncode;
pub struct UrlDecode;
pub struct UnicodeNames;
pub struct Colour;

impl FixedInt {
    pub fn get_base2_16_variant(value: &str) -> Option<Variant> {
        if let Some(variants) = Base2_16::identify(value) {

            if variants.len() == 1 {
                Some(variants.get(0).unwrap().clone())
            } else {
                let b10 = Variant("Base 10");

                if variants.contains(&b10) {
                    Some(b10)
                } else {
                    Some(variants.last().unwrap().clone())
                }
            }

        } else {
            None
        }
    }
}



impl Colour {
    ///Convert the 8-bit colour into 24-bit RGB
    pub fn _8_to_24(colour: u8) -> ansi_term::Colour {
        let blue = colour & 0x03;
        let green = (colour & 0x1C) >> 2;
        let red = (colour & 0xE0) >> 5;

        let red = Colour::scale(red, 8);
        let green = Colour::scale(green, 8);
        let blue = Colour::scale(blue, 4);

        ansi_term::Colour::RGB(red, green, blue)
    }

    ///Convert the 16-bit colour into 24-bit RGB
    pub fn _16_to_24(colour: u16) -> ansi_term::Colour {
        let blue = colour & 0x1f;
        let green = (colour & 0x7e0) >> 5;
        let red = (colour & 0xf800) >> 11;

        let red = Colour::scale(red as u8, 32);
        let green = Colour::scale(green as u8, 64);
        let blue = Colour::scale(blue as u8, 32);

        ansi_term::Colour::RGB(red, green, blue)
    }

    ///Scale up n bits of information to 8 bits
    fn scale(value: u8, total: u8) -> u8 {
        let step = 255.0  / (total as f32 - 1.0);

        (step * value as f32).round() as u8
    }
}

//

impl FixedFloat {
    fn mantissa_exponent_to_string(repr: (i8, u64, i16)) -> String {
        let (sign, mantissa, exponent) = repr;
        format!("Mantissa: {}({:b}), exponent: {}({:b}), sign: {}({:b})", mantissa, mantissa, exponent, exponent, sign, sign)
    }

    fn get_repr(bytes: &[u8]) -> (i8, u64, i16) {
        let int_bytes = match bytes.len() {
            2 => u16::from_le_bytes(bytes.try_into().unwrap()) as u64,
            4 => u32::from_le_bytes(bytes.try_into().unwrap()) as u64,
            8 => u64::from_le_bytes(bytes.try_into().unwrap()),
            _ => panic!("Invalid float ({} bytes long)", bytes.len()),
        };

        let (bits, mantissa_size, mantissa_mask, exponent_mask) = match bytes.len() {
            2 => (16, 10, 0x3FF, 0x7C00),
            4 => (32, 23, 0x07FFFFF, 0x7F800000),
            8 => (64, 52, 0xFFFFFFFFFFFFF, 0x7FF0000000000000),
            _ => panic!("Invalid float ({} bytes long)", bytes.len()),
        };

        let mantissa = mantissa_mask & int_bytes;

        let sign = if int_bytes >> (bits - 1) == 1 { -1 } else { 1 };

        let exponent = (exponent_mask & int_bytes) >> mantissa_size;

        (sign, mantissa, exponent as i16)
    }

    pub fn bytes_to_mes_string(bytes: &[u8]) -> String {
        Self::mantissa_exponent_to_string(Self::get_repr(bytes))
    }
}

impl Base2_16 {
    pub fn ascii_to_num(byte: u8) -> u8 {
        if byte >= '0' as u8 && byte <= '9' as u8 {
            byte - ('0' as u8)
        } else if byte >= 'a' as u8 && byte <= 'f' as u8 {
            byte - ('a' as u8) + 10
        } else {
            byte - ('A' as u8) + 10
        }
    }

    pub fn num_to_ascii(byte: u8, is_capital: bool) -> u8 {
        if byte <= 9 {
            byte + ('0' as u8)
        } else {
            if is_capital {
                byte - 10 + ('A' as u8)
            } else {
                byte - 10 + ('a' as u8)
            }
        }
    }

    pub fn is_hex(byte: u8) -> bool {
        byte >= '0' as u8 && byte <= '9' as u8 || byte >= 'a' as u8 && byte <= 'f' as u8 || byte >= 'A' as u8 && byte <= 'F' as u8
    }

    pub fn is_oct(byte: u8) -> bool {
        byte >= '0' as u8 && byte <= '7' as u8
    }

    pub fn get_base(variant: & Variant) -> u64 {
        u64::from_str(&(variant.0)[5..]).unwrap()
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Variant(pub & 'static str);


src/fromir.rs
use crate::common::{Variant, Colour};
use half::f16;
use std::convert::TryInto;
use std::str::{from_utf8, from_utf8_unchecked};
use std::net::{SocketAddrV4, SocketAddrV6};
use std::borrow::Cow;
use crate::endian::{Endianness};

use ansi_term::{ANSIGenericString, Style, Color};

//A trait that defines functions to convert from IR to T
pub trait FromIR {

    /// Used to identify all the possible variants that the IR can represent
    fn variants(ir: & [u8]) -> Option<Vec<Variant>>;

    ///Used to convert an IR to a String. NOTE: this function does NOT check to make sure that the IR can be converted, and will panic if the conversion fails
    fn encode(ir: & [u8], variant: Variant) -> ANSIGenericString<str>;
}

impl FromIR for crate::common::Base2_16 {
    fn variants(_ir: & [u8]) -> Option<Vec<Variant>> {
        //Any set of bytes can be converted to base 2-16
        Some(vec![Variant("Base 2"), //Variant("Base 3"), Variant("Base 4"), Variant("Base 5"), Variant("Base 6"), Variant("Base 7"),
             Variant("Base 8"), //Variant("Base 9"),
             Variant("Base 10"), //Variant("Base 11"), Variant("Base 12"), Variant("Base 13"), Variant("Base 14"), Variant("Base 15"),
             Variant("Base 16")])
    }

    fn encode(ir: & [u8], variant: Variant) -> ANSIGenericString<str> {
        let base = Self::get_base(&variant);

        let mut base_n_list = convert_base::Convert::new(256, base).convert::<u8, u8>(ir.as_ref());

        let mut string = String::new();

        for byte in base_n_list.iter_mut() {
            string.insert(0, Self::num_to_ascii(*byte, true) as char)
        }

        Style::default().paint(string)
    }
}

impl FromIR for crate::common::FixedFloat {
    fn variants(ir: & [u8]) -> Option<Vec<Variant>> {
        let len = ir.as_ref().len();

        let (mut variants, float) = match len {
            2 => (vec![Variant("16-bit"), Variant("16-bit mantissa/exponent")], f16::from_le_bytes(ir.as_ref().try_into().unwrap()).to_f64()),
            4 => (vec![Variant("32-bit"), Variant("32-bit mantissa/exponent")], f32::from_le_bytes(ir.as_ref().try_into().unwrap()) as f64),
            8 => (vec![Variant("64-bit"), Variant("64-bit mantissa/exponent")], f64::from_le_bytes(ir.as_ref().try_into().unwrap())),
            _ => return None,
        };

        if !float.is_infinite() && !float.is_nan() {
            match len {
                2 => variants.push(Variant("16-bit scientific notation")),
                4 => variants.push(Variant("32-bit scientific notation")),
                8 => variants.push(Variant("64-bit scientific notation")),
                _ => {}
            }


            if float.log10().abs() < 27.0 {
                match len {
                    2 => variants.push(Variant("16-bit SI notation")),
                    4 => variants.push(Variant("32-bit SI notation")),
                    8 => variants.push(Variant("64-bit SI notation")),
                    _ => {}
                }
            }
        }

        Some(variants)

    }

    fn encode(ir: & [u8], variant: Variant) -> ANSIGenericString<str> {


        let format = match &variant.0[6..] {
            "" => "f",
            " mantissa/exponent" => return Style::default().paint(Self::bytes_to_mes_string(ir)),
            " scientific notation" => "e",
            " SI notation" => "s",
            _ => panic!("Invalid variant in FromIT FixedFloat")
        };

        let num = format_num::NumberFormat::new();

        let number = match &variant.0[0..6] {
            "16-bit" => f16::from_le_bytes(ir.as_ref().try_into().unwrap()).to_f64(),
            "32-bit" => f32::from_le_bytes(ir.as_ref().try_into().unwrap()) as f64,
            "64-bit" => f64::from_le_bytes(ir.as_ref().try_into().unwrap()),
            _ => panic!("Invalid variant in FromIT FixedFloat")
        };

        Style::default().paint(num.format(format, number))
    }
}

impl FromIR for crate::common::FixedInt {
    fn variants(ir: & [u8]) -> Option<Vec<Variant>> {
        let len = ir.as_ref().len();

        match len {
            1 => Some(vec![Variant("u8"), Variant("i8")]),
            2 => Some(vec![Variant("u16"), Variant("i16")]),
            4 => Some(vec![Variant("u32"), Variant("i32")]),
            8 => Some(vec![Variant("u64"), Variant("i64")]),
            16 => Some(vec![Variant("u128"), Variant("i128")]),
            _ => None,
        }
    }

    fn encode(ir: & [u8], variant: Variant) -> ANSIGenericString<str> {
        Style::default().paint(match variant.0 {
            "u8" => u8::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "i8" => i8::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "u16" => u16::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "i16" => i16::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "u32" => u32::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "i32" => i32::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "u64" => u64::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "i64" => i64::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "u128" => u128::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            "i128" => i128::from_le_bytes(ir.as_ref().try_into().unwrap()).to_string(),
            _ => panic!("Invalid variant in FromIT FixedInt")
        })
    }
}

impl FromIR for crate::common::DateTime {
    fn variants(ir: & [u8]) -> Option<Vec<Variant>> {
        let len = ir.as_ref().len();

        match len {
            4 => Some(vec![Variant("32-bit rfc2822"), Variant("32-bit rfc3339")]),
            8 => {

                //Not all combinations of 64-bits result in a valid date, so we

                let le_timestamp = i64::from_le_bytes(ir.as_ref().try_into().unwrap());
                let be_timestamp = i64::from_be_bytes(ir.as_ref().try_into().unwrap());

                if chrono::NaiveDateTime::from_timestamp_opt(le_timestamp, 0).is_some() || chrono::NaiveDateTime::from_timestamp_opt(be_timestamp, 0).is_some() {
                    Some(vec![Variant("64-bit rfc2822"), Variant("64-bit rfc3339")])
                } else {
                    None
                }
            },
            _ => None,
        }
    }

    fn encode(ir: & [u8], variant: Variant) -> ANSIGenericString<str> {

        let timestamp = match &(variant.0)[0..2] {
            "64" => i64::from_le_bytes(ir.as_ref().try_into().unwrap()),
            "32" => i32::from_le_bytes(ir.as_ref().try_into().unwrap()) as i64,
            _ => panic!("Invalid variant (width) in FromIT DateTime"),
        };

        if let Some(datetime) = chrono::NaiveDateTime::from_timestamp_opt(timestamp, 0) {
            let datetime = chrono::DateTime::<chrono::Utc>::from_utc(datetime, chrono::Utc);

            Style::default().paint(match &(variant.0)[7..14] {
                "rfc2822" => datetime.to_rfc2822(),
                "rfc3339" => datetime.to_rfc3339(),
                "custom " => datetime.format(&(variant.0)[14..]).to_string(),
                _ => panic!("Invalid variant (format) in FromIT DateTime")
            })
        } else {
            Style::default().fg(Color::Red).paint("invalid or out-of-range datetime")
        }



    }
}

impl FromIR for crate::common::Unicode8 {
    fn variants(ir: &[u8]) -> Option<Vec<Variant>> {
        if from_utf8(ir).is_ok() {
            for character in unsafe {from_utf8_unchecked(ir).chars()} {
                if unicode_names2::name(character).is_none() {
                    return Some(vec![Variant("Literal String")])
                }
            }

            Some(vec![Variant("Literal String"), Variant("Unicode Names")])

        } else {
            None
        }
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {

        if variant.0 == "Literal String" {
            unsafe {
                Style::default().paint(String::from(from_utf8_unchecked(ir)))
            }
        } else if variant.0 == "Unicode Names" {
            let mut string = String::from("[");

            let unicode = unsafe {
                from_utf8_unchecked(ir)
            };

            for character in unicode.chars() {

                string.push_str(&unicode_names2::name(character).unwrap().to_string());
                string.push_str(" '");
                string.push(character);
                string.push_str("', ")
            }

            Style::default().paint(string)
        } else {
            panic!("Invalid variant in FromIR Unicode8");
        }
    }
}

impl FromIR for crate::common::IpV4 {
    fn variants(ir: &[u8]) -> Option<Vec<Variant>> {
        let len = ir.as_ref().len();

        match len {
            4 => Some(vec![Variant("without port")]),
            6 => Some(vec![Variant("with port")]),
            _ => None,
        }
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {

        let ip = std::net::Ipv4Addr::from(u32::from_le_bytes((&ir[0..4]).try_into().unwrap()));

        Style::default().paint(match variant.0 {
            "with port" => {
                let port = u16::from_le_bytes((&ir[4..6]).try_into().unwrap());

                SocketAddrV4::new(ip, port).to_string()
            },
            "without port" => {

                ip.to_string()
            }
            _ => panic!("Invalid variant in FromIT IpV4"),
        })
    }
}

impl FromIR for crate::common::IpV6 {
    fn variants(ir: &[u8]) -> Option<Vec<Variant>> {
        let len = ir.as_ref().len();

        match len {
            16 => Some(vec![Variant("without port")]),
            18 => Some(vec![Variant("with port")]),
            _ => None,
        }
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {

        let ip = std::net::Ipv6Addr::from(u128::from_le_bytes((&ir[0..16]).try_into().unwrap()));

        Style::default().paint(match variant.0 {
            "with port" => {
                let port = u16::from_le_bytes((&ir[16..18]).try_into().unwrap());

                SocketAddrV6::new(ip, port, 0, 0).to_string()
            },
            "without port" => {

                ip.to_string()
            }
            _ => panic!("Invalid variant in FromIT IpV4"),
        })
    }
}

impl FromIR for crate::common::Base91 {
    fn variants(_ir: &[u8]) -> Option<Vec<Variant>> {
        Some(vec![Variant("")])
    }

    fn encode(ir: &[u8], _variant: Variant) -> ANSIGenericString<str> {
        unsafe {
            Style::default().paint(String::from_utf8_unchecked(base91::slice_encode(ir)))
        }
    }
}

impl FromIR for crate::common::Base85 {
    fn variants(_ir: &[u8]) -> Option<Vec<Variant>> {
        Some(vec![Variant("z85"), Variant("ascii85")])
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {
        Style::default().paint(match variant.0 {
            "z85" => z85::encode(ir),
             "ascii85" => ascii85::encode(ir),
            _ => panic!("Invalid variant in FromIR Base85")
        })
    }
}

impl FromIR for crate::common::Base64 {
    fn variants(_ir: &[u8]) -> Option<Vec<Variant>> {
        Some(vec![Variant("Bcrypt"),
             Variant("BinHex"),
             Variant("crypt"),
             Variant("IMAP UTF-7"),
             Variant("Standard"),
             Variant("Standard no padding"),
             Variant("URL-safe"),
             Variant("URL-safe no padding"),])
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {
        Style::default().paint(match variant.0 {
            "Bcrypt" => base64::encode_config(ir, base64::BCRYPT),
            "BinHex" => base64::encode_config(ir, base64::BINHEX),
            "crypt" => base64::encode_config(ir, base64::CRYPT),
            "IMAP UTF-7" => base64::encode_config(ir, base64::IMAP_MUTF7),
            "Standard" => base64::encode_config(ir, base64::STANDARD),
            "Standard no padding" => base64::encode_config(ir, base64::STANDARD_NO_PAD),
            "URL-safe" => base64::encode_config(ir, base64::STANDARD_NO_PAD),
            "URL-safe no padding" => base64::encode_config(ir, base64::URL_SAFE_NO_PAD),
            _ => panic!("Invalid variant in FromIr Base64"),
        })
    }
}

impl FromIR for crate::common::ByteList {
    fn variants(_ir: &[u8]) -> Option<Vec<Variant>> {
        Some(vec![Variant("")])
    }

    fn encode(ir: &[u8], _variant: Variant) -> ANSIGenericString<str> {
        Style::default().paint(format!("({} byte(s)) {:?}", ir.len(), ir))
    }
}

impl FromIR for crate::common::UUID {
    fn variants(ir: &[u8]) -> Option<Vec<Variant>> {
        if uuid::Uuid::from_slice(ir).is_ok() {
            Some(vec![Variant("")])
        } else {
            None
        }
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {
        if variant.0 == "" {
            Style::default().paint(uuid::Uuid::from_slice(ir).unwrap().to_string())
        } else {
            panic!("Invalid variant in FromIT uuid")
        }
    }
}

impl FromIR for crate::common::EscapedString {
    fn variants(_ir: &[u8]) -> Option<Vec<Variant>> {
        Some(vec![Variant("")])
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {
        if variant.0 == "" {
            Style::default().paint(crate::escape::EscapeSequence::encode(ir))
        } else {
            panic!("Invalid variant in FromIR EscapedString");
        }
    }
}

impl FromIR for crate::common::UrlEncode {
    fn variants(_ir: &[u8]) -> Option<Vec<Variant>> {
        Some(vec![Variant("")])
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {
        if variant.0 == "" {
            Style::default().paint(urlencoding::encode_binary(ir).to_string())
        } else {
            panic!("Invalid variant in FromIR UrlEncode");
        }
    }
}

impl FromIR for crate::common::UrlDecode {
    fn variants(ir: &[u8]) -> Option<Vec<Variant>> {
        if from_utf8(ir).is_ok() {
            Some(vec![Variant("Legacy"), Variant("RFC 3986")])
        } else {
            None
        }
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {

        let string = if variant.0 == "RFC 3986" {
            let string = String::from_utf8_lossy(ir);
            Cow::from(string.replace("+", "%20"))
        } else {
            String::from_utf8_lossy(ir)
        };

        if variant.0 == "RFC 3986" || variant.0 == "Legacy" {
            Style::default().paint(urlencoding::decode(&string).unwrap())
        } else {
            panic!("Invalid variant in FromIR UrlDecode");
        }
    }
}

impl FromIR for crate::common::Colour {
    fn variants(ir: &[u8]) -> Option<Vec<Variant>> {
        let variants = match ir.len() {
            1 => vec![Variant("8-bit color"), Variant("8-bit greyscale"), Variant("8-bit terminal")],
            2 => vec![Variant("16-bit color")],
            3 => vec![Variant("24-bit color"), Variant("24-bit rgb"), Variant("24-bit hsl")],
            4 => vec![Variant("32-bit color")],
            _ => return None
        };

        Some(variants)
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {

        let colour = match variant.0 {
            "8-bit color" => {
                Colour::_8_to_24(*ir.get(0).unwrap())
            },
            "8-bit greyscale" => { let byte = *ir.get(0).unwrap(); Color::RGB(byte, byte, byte) },
            "8-bit terminal" => { Color::Fixed(*ir.get(0).unwrap()) },
            "16-bit color" => {Colour::_16_to_24(u16::from_le_bytes(ir.try_into().unwrap()))},
            "24-bit color" | "24-bit rgb" | "24-bit hsl" => {
                let red = *ir.get(2).unwrap();
                let green = *ir.get(1).unwrap();
                let blue = *ir.get(0).unwrap();

                Color::RGB(red, green, blue)
            },
            "32-bit color" => {
                let red = *ir.get(2).unwrap();
                let green = *ir.get(1).unwrap();
                let blue = *ir.get(0).unwrap();
                Color::RGB(red, green, blue)
            },
            _ => panic!("Invalid variant in FromIR Colour")
        };

        match variant.0 {
            "24-bit rgb" => {
                if let Color::RGB(r, g, b) = colour {
                    Style::default().paint(format!("rgb({}, {}, {})", r, g, b))
                } else {
                    panic!("Invalid variant in FromIR colour");
                }

            },
            "24-bit hsl" => {
                if let Color::RGB(r, g, b) = colour {

                    let slice = [b, g, r];

                    let hsl_value = hsl::HSL::from_rgb(&slice);

                    Style::default().paint(format!("hsl({}, {}, {})", hsl_value.h, hsl_value.s, hsl_value.l))
                } else {
                    panic!("Invalid variant in FromIR colour");
                }
            },
            _ => Style::default().fg(colour).paint("⬛"),
        }


    }
}

/*

impl FromIR for crate::common:: {
    fn variants(ir: &[u8]) -> Option<Vec<Variant>> {
        unimplemented!()
    }

    fn encode(ir: &[u8], variant: Variant) -> ANSIGenericString<str> {
        unimplemented!()
    }

    fn endianness() -> Endianness {
        Endianness::Default
    }
}

*/


